Issue#390 - bold commands in codeblocks (#413)
redhat-documentation/supplementary-style-guide: supplementary_style_guide/style_guidelines/formatting.adoc

[[formatting]]
= Formatting

[[commands-in-code-blocks]]
== Commands in code blocks

Use a single command per code block for each procedure step. This approach helps with readability and makes it possible for the copy button in code blocks to work correctly.

By default, use bold formatting for commands in code blocks to visually distinguish the commands from their lead-in sentences and from the command prompts and example outputs. For consistency, use bold formatting for commands even when no command output is shown in the code block.

[NOTE]
====
To apply formatting in a code block, you must use the `quotes` link:https://docs.asciidoctor.org/asciidoc/latest/subs/apply-subs-to-blocks/[AsciiDoc substitution].
====

.Example AsciiDoc: A command and its output in a code block

>   Verify that the `libvirt` default network is active and configured to start automatically:

Reviewer: My only feedback is whether this example should actually be shown as a bullet or numbered step. But I don't think that's important enough to hold up the PR over since it has 2 other approvals. Just something to consider if we should adjust in a separate PR or not!

[[formatting]]
= Formatting

[[commands-in-code-blocks]]
== Commands in code blocks

> Use a single command per code block per procedure step. This helps with readability and makes it possible for the copy button in code blocks to work correctly.

Reviewer 2: ```suggestion
Use a single command per code block for each procedure step. This helps with readability and makes it possible for the copy button in code blocks to work correctly.
```

> As a default, use bold formatting for commands in code blocks, to visually distinguish them from their lead-in sentences and example outputs. For consistency, do this even when no output is shown in the code block.

Reviewer 2: ```suggestion
As a default, use bold formatting for commands in code blocks to visually distinguish them from their lead-in sentences and example outputs. For consistency, do this even when no output is shown in the code block.
```

[NOTE]
====
To apply formatting in a code block, you must use the `quotes` link:https://docs.asciidoctor.org/asciidoc/latest/subs/apply-subs-to-blocks/[AsciiDoc substitution].
====

.Example AsciiDoc: A command and its output in a code block

  Verify that the `libvirt` default network is active and configured to start automatically:
  
  [subs="+quotes"]
  ----
  # *virsh net-list --all*
  Name      State    Autostart   Persistent
  --------------------------------------------
  default   active   yes         yes
  ----

This example renders as follows in HTML:

====
Verify that the `libvirt` default network is active and configured to start automatically:

[subs="+quotes"]
----
# *virsh net-list --all*
Name      State    Autostart   Persistent
--------------------------------------------
default   active   yes         yes
----
====

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Author: I don't think this is good guidance. Writers should still be able to use bolded strings in outputs if necessary, especially if in the given code block, the output is more important than the command itself. Also, it would be rather strange to express the same kind of emphasis differently in the very same code block.

Your other suggestions are decidedly improvements, though, thank you :-)

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Reviewer 3: @jherrman Apologies for the delay in replying to your comment. It seems odd to me to say bold the command in one place and not in another, but perhaps I misunderstood the intent of this update.

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Author: Well, the idea is basically "bold the important part, so that it's visually distinct" - and since the command itself is usually the important part of a CLI example (and the given step in general), it would be the default part do bold. However,  when a part of the output is as important as the command in the given context, or more, we should not bar the writer from bolding something else than the command, and potentially even leaving the command un-bolded.

It's entirely possible I did not formulate that clearly enough in the actual source update, though...

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Reviewer 2: @jafiala Can you give us an example of what you're thinking of with bolding pieces of the output?

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Author: I suspect that @ was meant to aim at me :-) A few examples I fished out:

- https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html-single/configuring_and_managing_virtualization/index#setting-up-secure-execution-on-ibm-z_securing-virtual-machines-in-rhel -> Verification (highlights where in the output the user can find the line they should look for)
- https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html-single/configuring_and_managing_virtualization/index#setting-up-secure-execution-on-ibm-z_securing-virtual-machines-in-rhel -> Prerequisites and Step 8i (again, highlighting what the user should look for)

I probably don't have examples with bolded output and un-bolded commands, but if in the context you really need to highlight a part of the output rather than the command  you use to get it, I don't think we should prohibit that, either.

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Reviewer 3: Thanks for the examples, @jherrman. 

I think it's better to highlight such lines in the command output by using a callout instead of bold text? The following example shows callouts (in a code example, but the same method could be used for command output):
https://docs.openshift.com/pipelines/1.13/about/understanding-openshift-pipelines.html

WDYT?

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Author: Honestly, callouts do not seem like an optimal solution to me.

- They are pretty wonky to work with in adoc (for instance can only be at the end of a line, cannot be used out of sequence, one callout cannot be used repeatedly, and they must always have an explanation under the codeblock, AFAIK). Also, Customer Portal stylesheets do not render them very well, currently.

- When they work as expected, callouts are visually very distinct, which may not always be what you want when highlighting a small part of a command output. 

If the general intention behind this update is "bold out the operative/important parts of the code block for greater visual clarity", I would say the output should follow that consistently with the commands. Also, it's easier for writers to internalize and follow.

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Reviewer 3: @jherrman and @dfitzmau Thank you both! I understand that callouts might not be optimal, but I wonder if we could highlight the command output in some other way rather than by reusing the bold format. 

Do you know whether the highlight syntax described [here](https://docs.asciidoctor.org/asciidoc/latest/text/highlight/) would render correctly in Customer Portal and OpenShift docs?

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Reviewer 4: I tried that syntax a few years ago on the CP and it did render, but I was informed to drop it as the yellow highlighting was really distracting. It definetly grabs attention, but it distracts from the surrounding content. I can understand why yellow is a well-adopted colour for safety :yellow_heart:

> Do not use bold formatting for the output, unless you need to highlight a specific part of it. When you do emphasize a non-command item in a code block, you are not required to bold the command.

Author: I tested out the `#highlight#` markup earlier today. Renders reasonably well in asciidoctor (though, as Darragh points out, the colour is a bit much), but both bccutil and the Customer Portal stylesheets completely ignore it, unfortunately: 

https://dxp-docp-prod.apps.ext-waf.spoke.prod.us-west-2.aws.paas.redhat.com/documentation/en-us/zen_and_zany_asciidoc_training_titles/42/html-single/davozenis_advanced_asciidoc_training/index?lb_target=stage#proc_test-two_assembly_nested-test

Source:

> [subs="+quotes"]
> ----
> # *Sample command sting*
>
> Sample #output# string
----

> For both commands and outputs in code blocks, observe the correct mark-up for user-replaced values, as described in xref:user-replaced-values[] and xref:user-replaced-values-xml[].

Reviewer 2: ```suggestion
For both commands and outputs in code blocks, observe the correct markup for user-replaced values, as described in xref:user-replaced-values[] and xref:user-replaced-values-xml[].
```

* _Configuring the node port service range_
* _How to perform an unsupported conversion from a RHEL-derived Linux distribution to RHEL_

> [[commands-in-code-blocks]]

Author: You're right, that's a topic that falls quite squarely under "Commands in code blocks]", so I'll add a mention of that, too.

> == Commands in code blocks

Reviewer 4: +1 For Technical examples

> == Commands in code blocks

Author: Hm, I'd say it fits rather well under the "Formatting" section, since "Technical examples" primarily seems to deal with guidelines for the content itself rather than its markup. 

Also, the "User-replaced values" parts, which deal with a very much related topic, including code block content  (which is why I initially bunched up the new section with them), are under Formatting as well.

As a default, use bold formatting for commands in code blocks, to visually distinguish them from their lead-in sentences and example outputs.

> 

Author: I think so, good point.

.Example AsciiDoc: Command and its output in a code block

>   Verify that the `libvirt` default network is active and configured to start automatically:

Author: Actually, I just tested this and it seems to break the formatting of the example in the render for some reason. So, given that it's not really important for the example lead-in sentence to be a step, I'd leave it out.

  +
  [subs="+quotes"]
  ----
  # *virsh net-list --all*
  Name      State    Autostart   Persistent
  --------------------------------------------
  default   active   yes         yes
  ----

> This renders as:

Reviewer 4: Incomplete sentence. How about: `The previous AsciiDoc example outputs as follows`?

I'm suggesting this in the event that a new writer scans the RH SSG (to validate term usability) to check for instances of "This renders as" and they then apply it to customer docs.

Author: This exact wording was actually used in multiple other examples in the section, and I just copied it as a "template". Doesn't change the fact that it's an incomplete sentence, so I'll update the other occurrences, too.

====
Verify that the `libvirt` default network is active and configured to start automatically:

[subs="+quotes"]
----
# *virsh net-list --all*
Name      State    Autostart   Persistent
--------------------------------------------
default   active   yes         yes
----
====

> Disregard this guideline for example when the command is not an important part of what you want to show in the given code block.

Author: Well, I do want to leave people some room to manoeuvre around this if they feel so inclined. Your suggestion for the wording does make sense, though.

> In addition, both in commands and outputs in code blocks, observe the correct mark-up for user-replaced values, as described in xref:user-replaced-values[] and xref:user-replaced-values-xml[].

Reviewer 4: The three "in" instances makes me stop and reread the sentence. How about:

> For a command and its output in a code block, observe the correct mark-up for user-replaced values, as described in xref:user-replaced-values[] and xref:user-replaced-values-xml[].

  [subs="+quotes"]
  ----

>   $ *oc describe node __<node_name>__*

Reviewer 4: The division from the lead-in sentence is marked by the code block's shaded box, so I'm not really sure bolding the command is necessary. I understand bolding a command to distinguish it from command output that resides in the same code block.

Author: Well, the original pitch for this change partly stemmed from the fact that the light-grey shading of the code block in the current stylesheets does not really do much to highlight whatever it contains. 

In addition, for the sake of consistency, we agreed that commands in code blocks should be bolded regardless of whether they are accompanied by the output or not.

Reviewer 4: Thanks for the reply. Phew, a lot of code blocks will need to be updates after this update!  Would detailing the "quotes" subs attribute be important so that a writer knows the importance of the attribute on how the AsciiDoc processor renders asterisk symbol as bold highlighting?

A link could be provided that points to https://docs.asciidoctor.org/asciidoc/latest/subs/apply-subs-to-blocks/ . I see a few links to Asciidoctor docs in the RH SSG.
